src/Qir/Runtime/lib/QSharpCore/qsharp__core__qis.hpp

```diff
@@ -7,6 +7,20 @@
 
 struct QirArray;
 
+struct QirRTuple
+{
+    PauliId pauli;
+    double angle;
+    QUBIT* target;
+};
+
+struct QirExpTuple // NOLINT
+{
+    QirArray* paulis;
+    double angle;
+    QirArray* targets;
+};
```

Comment:
I assume all these structs need to be in the "extern C" section.

Reply:
I'm actually not sure on this one... on the one hand `QirArray` and `QirString` are not extern C because they are not expected to be used directly by C code or QIR but rather passed around as opaque pointers. In this case, these tuple structs are convenience definitions for what is ultimately a capture tuple, which in QIR will be just an LLVM struct. That seems to imply that it would be the responsibility of every implementer to define a struct that matches their implementation of these tuple contents (or dynamically interpret the memory in the function body). So I don't expect any extern C callers to reuse these and would instead use their own struct definitions or make use of the `__quantum__rt__tuple_create` and related runtime features. But I'm not really sure what the right behavior here is.

Reply:
I feel that eventually we will follow the item [Avoid Exposing Entities With Multiple Layouts](https://github.com/microsoft/qsharp-runtime/blob/main/src/Qir/QIR-RT-API-Design-Guidelines.md#avoid-exposing-entities-with-multiple-layouts). No need to follow that item in this PR (feel free to leave as is), but if you like you can try to make a first step in that direction and "test that item" :-).

Reply:
That's a really good point, thanks! I think adjusting our use of tuples here (and in other places where argument tuples or capture tuples are used) to properly follow the guidance you put together for QIR APIs will be our best bet for success.